File: Go/1679. Max Number of K-Sum Pairs.go
package main

//goland:noinspection GoUnusedFunction
func maxOperations(nums []int, k int) int {
	numsCount := make(map[int]int)

	for _, num := range nums {
		numsCount[num]++
	}

	operations := 0

	for num, count := range numsCount {
		if indexCount, ok := numsCount[k-num]; ok {
			operations += min(count, indexCount)
		}
	}

	return operations / 2
}

File: TypeScript/1679. Max Number of K-Sum Pairs.ts
// noinspection JSUnusedGlobalSymbols

const maxOperations = (nums: number[], k: number): number => {
  const numCount = new Map<number, number>()

  for (const num of nums) {
    numCount.set(num, (numCount.get(num) ?? 0) + 1)
  }

  let operations = 0

  for (const [num, count] of numCount.entries()) {
    if (numCount.has(k - num)) {
      operations += Math.min(count, numCount.get(k - num) ?? 0)
    }
  }

  return Math.floor(operations / 2)
}
